core/src/main/java/org/bitcoinj/core/Utils.java

```diff
@@ -309,13 +313,18 @@ public static BigInteger decodeMPI(byte[] mpi, boolean hasLength) {
      * Thus, all the complexities of the sign bit and using base 256 are probably an implementation accident.</p>
      */
     public static BigInteger decodeCompactBits(long compact) {
+        if (prevCompact == compact) {
```

Comment:
This is not thread safe. Caching to avoid complicated computation should use a direct long->BigInteger mapping. Some sort of LRU Map.